programs/dynamic-bonding-curve/src/state/virtual_pool.rs

```diff
@@ -276,101 +426,290 @@ impl VirtualPool {
+                    total_input_amount = total_input_amount.safe_add(input_amount)?;
                     current_sqrt_price = next_sqrt_price;
                     amount_left = amount_left.safe_sub(
-                        max_amount_in
+                        max_amount_out
                             .try_into()
                             .map_err(|_| PoolError::TypeCastFailed)?,
                     )?;
                 }
             }
         }
         if amount_left != 0 {
-            let next_sqrt_price = get_next_sqrt_price_from_input(
+            let next_sqrt_price = get_next_sqrt_price_from_output(
                 current_sqrt_price,
                 config.curve[0].liquidity,
                 amount_left,
                 true,
             )?;
 
-            let output_amount = get_delta_amount_quote_unsigned(
+            let input_amount = get_delta_amount_base_unsigned(
                 next_sqrt_price,
                 current_sqrt_price,
                 config.curve[0].liquidity,
+                Rounding::Up,
+            )?;
+            total_input_amount = total_input_amount.safe_add(input_amount)?;
+            current_sqrt_price = next_sqrt_price;
+        }
+
+        Ok(SwapAmount {
+            amount_in: total_input_amount,
+            output_amount: amount_out,
+            next_sqrt_price: current_sqrt_price,
+        })
+    }
+
+    fn get_swap_out_amount_from_quote_to_base(
+        &self,
+        config: &PoolConfig,
+        amount_out: u64,
+    ) -> Result<SwapAmount> {
+        let mut total_input_amount = 0u64;
+        let mut current_sqrt_price = self.sqrt_price;
+        let mut amount_left = amount_out;
+
+        for i in 0..config.curve.len() {
+            if config.curve[i].sqrt_price == 0 || config.curve[i].liquidity == 0 {
+                break;
+            }
+            if config.curve[i].sqrt_price > current_sqrt_price {
+                let max_amount_out = get_delta_amount_base_unsigned_256(
+                    current_sqrt_price,
+                    config.curve[i].sqrt_price,
+                    config.curve[i].liquidity,
+                    Rounding::Down,
+                )?;
+                if U256::from(amount_left) < max_amount_out {
+                    let next_sqrt_price = get_next_sqrt_price_from_output(
+                        current_sqrt_price,
+                        config.curve[i].liquidity,
+                        amount_left,
+                        false,
+                    )?;
+
+                    let input_amount = get_delta_amount_quote_unsigned(
+                        current_sqrt_price,
+                        next_sqrt_price,
+                        config.curve[i].liquidity,
+                        Rounding::Up,
+                    )?;
+                    total_input_amount = total_input_amount.safe_add(input_amount)?;
+                    current_sqrt_price = next_sqrt_price;
+                    break;
```

Comment:
we need to set amount_left = 0, and validate amount_left is zero when we are out of loop?